### PR TITLE
IGNITE-20640 Fix raft clients after stable assignments update

### PR DESCRIPTION
**Pull request by kgusakov**
https://issues.apache.org/jira/browse/IGNITE-20640

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java [hunk 1]
```diff
@@ -482,6 +482,13 @@ public ClusterService clusterService() {
         return cluster;
     }
 
+    @Override
+    public void updateConfiguration(PeersAndLearners configuration) {
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java:486]: As @kgusakov mentioned given methods aren't thread safe. Several tickets will be created in order to fix this.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 2]
```diff
@@ -583,6 +585,59 @@ void testRebalanceWithTheSameNodes() throws Exception {
...
+    void testRaftClientsUpdatesAfterRebalance() throws InterruptedException, ExecutionException, TimeoutException {
+        Node node = getNode(0);
+
+        createZone(node, ZONE_NAME, 1, 1);
+
+        createTable(node, ZONE_NAME, TABLE_NAME);
+
+        assertTrue(waitForCondition(() -> getPartitionClusterNodes(node, 0).size() == 1, AWAIT_TIMEOUT_MILLIS));
+
+        Set<Assignment> assignedNodeBeforeRebalance = getPartitionClusterNodes(node, 0);
```
Comment (sanpwc) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:598]: Minor. It's probably a typo. assigned -> assignments in assignedNodeBeforeRebalance

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 3]
```diff
@@ -2157,6 +2157,16 @@ protected CompletableFuture<Void> handleChangeStableAssignmentEvent(WatchEvent e
 
         return metaStorageMgr.get(pendingPartAssignmentsKey(tablePartitionId), stableAssignmentsWatchEvent.revision())
                 .thenComposeAsync(pendingAssignmentsEntry -> {
+                    CompletableFuture<Void> fut = completedFuture(null);
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2160]: One more minor. I'd also give it more meaningful name.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 4]
```diff
@@ -2157,6 +2157,16 @@ protected CompletableFuture<Void> handleChangeStableAssignmentEvent(WatchEvent e
 
         return metaStorageMgr.get(pendingPartAssignmentsKey(tablePartitionId), stableAssignmentsWatchEvent.revision())
                 .thenComposeAsync(pendingAssignmentsEntry -> {
+                    CompletableFuture<Void> raftClientUpdateFuture = completedFuture(null);
+
+                    if (!stableAssignments.equals(ByteUtils.fromBytes(evt.entryEvent().oldEntry().value()))) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2162]: What is this condition for? How can new assignments be equal to previous assignments?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 5]
```diff
@@ -2157,6 +2157,16 @@ protected CompletableFuture<Void> handleChangeStableAssignmentEvent(WatchEvent e
...
+                        raftClientUpdateFuture = tablesById(evt.revision()).thenAccept(t -> {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2163]: Please add some comments about the purpose of this code

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 6]
```diff
@@ -587,6 +587,59 @@ void testRebalanceWithTheSameNodes() throws Exception {
...
+
+        createZone(node, ZONE_NAME, 1, 1);
+
+        createTable(node, ZONE_NAME, TABLE_NAME);
+
+        assertTrue(waitForCondition(() -> getPartitionClusterNodes(node, 0).size() == 1, AWAIT_TIMEOUT_MILLIS));
+
+        Set<Assignment> assignmentsBeforeRebalance = getPartitionClusterNodes(node, 0);
+
+        Node newNodeForAssignment = nodes.stream().filter(n ->
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:602]: May I propose the following refactoring to make the code more readable:
```
String newNodeNameForAssignment = nodes.stream()
        .map(n -> Assignment.forPeer(n.clusterService.nodeName()))
        .filter(assignment -> !assignmentsBeforeRebalance.contains(assignment))
        .findFirst()
        .orElseThrow()
        .consistentId();   
```

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 7]
```diff
@@ -587,6 +587,59 @@ void testRebalanceWithTheSameNodes() throws Exception {
...
+                !assignmentsBeforeRebalance.contains(Assignment.forPeer(n.clusterService.nodeName()))).findFirst().get();
+
+        Set<Assignment> newAssignment = Set.of(Assignment.forPeer(newNodeForAssignment.clusterService.nodeName()));
+
+        // Write the new assignments to metastore as a pending assignments.
+        {
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:608]: Why do you need these braces?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 8]
```diff
@@ -587,6 +587,59 @@ void testRebalanceWithTheSameNodes() throws Exception {
...
+            TablePartitionId partId = new TablePartitionId(getTableId(node, TABLE_NAME), 0);
+
+            ByteArray partAssignmentsPendingKey = pendingPartAssignmentsKey(partId);
+
+            Map<ByteArray, byte[]> msEntries = new HashMap<>();
+
+            byte[] bytesPendingAssignments = ByteUtils.toBytes(newAssignment);
+
+            msEntries.put(partAssignmentsPendingKey, bytesPendingAssignments);
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:617]: You can use `Map.of` here

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java (hunk 1 above)
Comment (kgusakov) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java:486]: Actually I meant that many methods from `RaftGroupServiceImpl` are not threadsafe, for example refreshMembers called in paralell with refreshMembers/addLearners/addPeers and other - can produce the client with the (peers, learners) pairs which is meaningless and weren't exist in any point of time in fact.

So, it is not connected with only this method, but I created the https://issues.apache.org/jira/browse/IGNITE-20738 and will mark this method with it.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java (hunk 7 above)
Comment (kgusakov) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:608]: The cheapest way to create the separate block with finite logic and linked comment. I use it in tests sometimes to increase the readability and to not spam the test class with one-shot methods. If this practice looks so extraordinary, I can remove it

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java (hunk 1 above)
Comment (sashapolo) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java:486]: You can mark the whole class with this

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java (hunk 7 above)
Comment (sashapolo) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:608]: it's ok if you like it, just a matter of taste

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java (hunk 8 above)
Comment (kgusakov) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java]: Thanks!

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 5 above)
Comment (kgusakov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java]: Done

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 9]
```diff
@@ -2157,6 +2157,16 @@ protected CompletableFuture<Void> handleChangeStableAssignmentEvent(WatchEvent e
 
         return metaStorageMgr.get(pendingPartAssignmentsKey(tablePartitionId), stableAssignmentsWatchEvent.revision())
                 .thenComposeAsync(pendingAssignmentsEntry -> {
+                    CompletableFuture<Void> raftClientUpdateFuture = completedFuture(null);
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java]: Thanks

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java (hunk 1 above)
Comment (kgusakov) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java:488]: Marked the whole class, yep

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 10]
```diff
@@ -587,6 +587,61 @@ void testRebalanceWithTheSameNodes() throws Exception {
...
+        // Write the new assignments to metastore as a pending assignments.
+        {
+            TablePartitionId partId = new TablePartitionId(getTableId(node, TABLE_NAME), 0);
+
+            ByteArray partAssignmentsPendingKey = pendingPartAssignmentsKey(partId);
+
+            byte[] bytesPendingAssignments = ByteUtils.toBytes(newAssignment);
+
+            node.metaStorageManager
+                    .putAll(Map.of(partAssignmentsPendingKey, bytesPendingAssignments))
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:620]: actually, why are we using `putAll` to insert one key? Looks like a simple `put` is enough =)

**Review by sashapolo — APPROVED** [review]
I've approved the PR with a minor comment

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java (hunk 10 above)
Comment (kgusakov) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java]: Artifact of the previous code states :) fixed, thanks

**Review by sanpwc — APPROVED** [review]
